woocommerce/woocommerce-gutenberg-products-block
# Enhance checkout filter system.

This PR is a follow up to https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/3750.

### Error messages on the validation
instead of skipping bad filters, we're going to throw a specific error for each error (not a string, doesn't contain label...).
Error being thrown would be caught by the filter system and highlighted to admins or skipped for users, just like other exceptions.
Error messages are now living in `util/validation`.

### Refactor passed in arguments
We now pass extensions as a top level prop to the filter, this is to align with the rest of our APIs like SlotFills and to make it easier for plugin author to access extended data.
I also renamed some arguments like blocks to be context and mixed usage of cartItem|lineItem to lineItem.

### Performance guards
We previously called all registered filters each the component did a recalculation (about 8-10 times on a single render), this is risky because we're calling 3PD code each time and we don't know how expensive it can be, therefore, I memozied all filters calls.
I tried to do that inside the filter directly (by making it a hook) but it didn't really work, probably because the passed in arguments are too complex.

### Cherry pick lost commits
Previous PRs had a couple of commits from @Aljullu that never made it to the merged PR for some reason, I was still able to see them, so I cherry picked them into this PR.

## Testing
PR needs 3971-gh-woocommerce/woocommerce-subscriptions in WCS.
Each step is a commit, starting from `throw errors on validation` commit.
- Make sure all filters are working fine, add subscription products to your cart and make sure "total due today" is showing, "x every y" and "due today x" are showing on each price, subtotal, and saving badge,
- Try returning a wrong value in WCS, if you're an admin, you should see the error, if you're a customer, you should only see a console error.
